### ISSUE TITLE
Directly using Nginx as a server without Uvicorn

### ISSUE THREAD

**Issue by rafsaf** [tiangolo/fastapi#3128]
Hi, as I understand it, I can implement fastapi application with (or at least I know such options):

- only Uvicorn 
- Uvicorn behind Nginx
- Uvicorn behind Gunicorn
- (Uvicorn behind something else?)

I found out that one can use Nginx unit directly with FastAPI as described [here in nginx unit docs](https://unit.nginx.org/howto/fastapi/). It works as expected. It seems that using only Nginx is slightly simpler for me (maybe somehow faster or slower too? why?) than Uvicorn + something, but I haven't found anything in docs/issues about using it directly and using Uvicorn alone isn't recommended in some cases (as far as i know), hence the questions.

**The questions are** 
- what are the possible drawbacks to using this approach / why should I avoid using Nginx directly 
- if there are no contraindications, what additional steps should I eventually take to make it work well
- what are the main differences between these solutions (Uvicorn vs Uvicorn + something vs Nginx only)

To provide any context on how I exactly implemented it with Nginx only, let's say in our application's root folder we have `config.json` in `config` folder and hello world fastapi app in `app/main.py`, then it might look like this:

**Dockerfile**
```dockerfile
FROM nginx/unit:1.23.0-python3.9
COPY ./config/config.json /docker-entrypoint.d/config.json
RUN mkdir build
COPY . ./build
RUN apt update && apt install -y python3-pip                                  \
    && pip3 install -r /build/requirements.txt                               \
    && apt remove -y python3-pip                                              \
    && apt autoremove --purge -y                                              \
    && rm -rf /var/lib/apt/lists/* /etc/apt/sources.list.d/*.list
EXPOSE 80
```
**config.json**
```json
{
    "listeners": {
        "*:80": {
            "pass": "applications/fastapi"
        }
    },

    "applications": {
        "fastapi": {
            "type": "python 3.9",
            "path": "/build/app/",
            "module": "main",
            "callable": "app"
        }
    }
}
```

**Comment by frankie567** [comment]
I wasn't aware of Nginx Unit, that's interesting!

First of all, let's clarify things a bit:

* FastAPI (thanks to its underlying server layer [Starlette](https://www.starlette.io/)) implement the [Python ASGI protocol](https://asgi.readthedocs.io/en/latest/). It's a standard to build web application with async support. It's the successor of WSGI, which is the standard protocol to build synchronous web applications.
* Uvicorn is an implementation of an ASGI server. Its role is to run ASGI apps (like FastAPI) and expose an HTTP server.
  * There are others ASGI server implementations out there, like Daphne or Hypercorn.
* Gunicorn, at its base, is a WSGI server implementation, widely used with Flask or Django for example.
  * However, it has interesting features to manage and monitor server processes, which may make it [more resilient in production](https://www.uvicorn.org/deployment/#using-a-process-manager).
  * That's why it's recommended to run FastAPI apps with Gunicorn configured with an Uvicorn worker, to make it work with ASGI protocol.
* I'm less aware of the implications of running Uvicorn behind Nginx, or even Uvicorn-Gunicorn-Nginx. I guess it allows you to have interesting rules to optimize static file serving or caching. I would say that Uvicorn-Gunicorn-Nginx may be a bit useless.

Now, for "Nginx Unit". From what I understand, it's an Nginx module implementing its own version of ASGI; so it comes "instead of" Uvicorn, it's not just Nginx on its own.

I see that they have an [implementation of ASGI in pure C](https://github.com/nginx/unit/blob/master/src/python/nxt_python_asgi.c). I would say that it should work quite well, after all, Nginx has proven to be *very* efficient for years, so I guess the people behind know their stuff.

The difference could be in terms of performance; and the only way to find out which one is better is to perform benchmarks. I didn't find existing for ASGI, that could be interesting to do.

**Comment by Kludex** [comment]
> The difference could be in terms of performance; and the only way to find out which one is better is to perform benchmarks. I didn't find existing for ASGI, that could be interesting to do.

You can try changing the event loop and see how it behaves:  
```python
import uvloop

asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
```

**Comment by Faylixe** [comment]
Wow, I wasn't aware of that either, that would be great to benchmark all existing process manager solution indeed, I personally ended up using Hypercorn with `uvloop` which is actually blazing fast and support heavy load (and also this is actually the only solution that is supporting HTTP/2).

**Comment by raphaelauv** [comment]
@Faylixe have a fair point HTTP/2 is not yet part of nginx unit https://github.com/nginx/unit/issues/388

**Comment by synodriver** [comment]
I have tested nginx unit using wrk. It's around 20% faster than uvicorn with a real application that queries sqlite, and 40% faster than hypercorn under the same condition. Due to it's high throughput, it may be the future of asgi server. As for http2, uvicorn does not support it either. Hypercorn do. So if you need extreme fast servers you can try nginx unit, but if you want more asgi's feature you may try hypercorn

**Comment by Kludex** [comment]
Can I see the benchmarks?

**Comment by synodriver** [comment]
Sure
![F9677C51-A26D-4ED4-A2D9-F043331F67E7](https://user-images.githubusercontent.com/50922489/134885748-014857de-301c-4040-b0d5-f13d27d15344.png)
![7F2994A1-13D1-4C4C-95A4-20A21B602055](https://user-images.githubusercontent.com/50922489/134885783-5ccf5735-b433-43f5-a368-7599535ebdd2.png)
![B991BFD9-BE6E-48AC-B5D1-C276DBB5BC51](https://user-images.githubusercontent.com/50922489/134885795-b65e6453-8ab7-4d7c-b767-f7d415ad2911.png)

**Comment by Kludex** [comment]
I mean the setup, which commands did you use, on what machines, what endpoints, etc...

**Comment by synodriver** [comment]
NGINX unit keeps high throughput under heavy load.

**Comment by synodriver** [comment]
> I mean the setup, which commands did you use, on what machines, what endpoints, etc...

I can create a new repo later...

**Comment by omidraha** [comment]
Related links:

https://unit.nginx.org/

https://fastapi.tiangolo.com/advanced/behind-a-proxy/

https://fastapi.tiangolo.com/deployment/manually/

https://www.uvicorn.org/#running-with-gunicorn

https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker

https://stackoverflow.com/a/63427961

https://stackoverflow.com/a/66426744

https://levelup.gitconnected.com/deploying-an-asynchronous-fastapi-on-nginx-unit-b038288bec5

https://dev.to/shuv1824/deploy-fastapi-application-on-ubuntu-with-nginx-gunicorn-and-uvicorn-3mbl

**Comment by rafsaf** [comment]
I don't see how it's related to either my original question (in which, by the way, I did provide an example of sample implementation using Nginx Unit working just fine even after few months and a link to docs - nobody really asked for a direct help or examples but clarification of things)

What about benchmarks - I would also love to see the whole setup of @synodriver (or made by anyone). Stuff like that even with careful design may have little to do with reality, there are just so many config options (of servers), VM implementation details, FastAPI details to consider (sync/async endpoins for example) etc.

**Comment by synodriver** [comment]
> I don't see how it's related to either my original question (in which, by the way, I did provide an example of sample implementation using Nginx Unit working just fine even after few months and a link to docs - nobody really asked for a direct help or examples but clarification of things) or mentioned benchmarks.
> 
> What about benchmarks - I would also love to see the whole setup of @synodriver (or made by anyone). Stuff like that even with careful design may have little to do with reality, there are just so many config options (of servers), VM implementation details, FastAPI details to consider (sync/async endpoins for example) etc.

Hi, [here](https://github.com/synodriver/asgi-server-benchmark) is the repo for bench mark.

**Comment by ghandic** [comment]
Thanks for posting your research code up @synodriver ! Could you add a raw md table to your repo README for direct comparison?

**Comment by Kludex** [comment]
Thanks! @synodriver

**Comment by blueray453** [comment]
[NGINX Unit documentation](https://unit.nginx.org/howto/fastapi/) do not use Unicorn. Is it ok to use this method?

**Comment by rafsaf** [comment]
Yes. Nginx Unit works just fine with FastAPI (I use it for Django apps too), you don't need unicorn or anything to run over http, just copy paste config code from docs you have linked.

But, but, you will probably need https anyway, I personally prefer some proxy, I didn't even try Nginx Unit "way" of applying tls but there is one apparently. Nginx Ingress, traefik for k8s, nginx proxy manager for swarm, alone traefik docker image or full nginx, probably tens of other solutions.

Gunicorn will be also fine tool for a job if you like it.

**Comment by blueray453** [comment]
I asked a question on [stackoverflow](https://stackoverflow.com/questions/69642063/after-nginx-unit-do-we-still-need-asgi-like-uvicorn-hypercorn-daphne-etc). There they said:

> nginx unit is an ASGI compliant server; you're replacing uvicorn/hypercorn/daphne/etc. with nginx unit instead. It's a choice you make just like you'd make using either of those three or other alternatives.

I am not understanding what it means. Does it mean, I have to install both nginx and nginx unit where nginx unit is the replacement for uvicorn / hypercorn / daphne etc. Or, nginx unit alone can do what nginx + uvicorn / hypercorn / daphne etc. does.
